bgpd/bgp_packet.c

```diff
@@ -2134,6 +2212,113 @@ static int bgp_route_refresh_receive(struct peer *peer, bgp_size_t size)
 				   SUBGRP_STATUS_DEFAULT_ORIGINATE);
 	}
 
+	if (subtype == BGP_ROUTE_REFRESH_BORR) {
+		/* A BGP speaker that has received the Graceful Restart
+		 * Capability from its neighbor MUST ignore any BoRRs for
+		 * an <AFI, SAFI> from the neighbor before the speaker
+		 * receives the EoR for the given <AFI, SAFI> from the
+		 * neighbor.
+		 */
+		if (CHECK_FLAG(peer->cap, PEER_CAP_RESTART_RCV)
+		    && !CHECK_FLAG(peer->af_sflags[afi][safi],
+				   PEER_STATUS_EOR_RECEIVED)) {
+			zlog_info(
+				"%s: rcvd route-refresh (BoRR) for %s/%s before EoR",
+				peer->host, afi2str(afi), safi2str(safi));
+			return BGP_PACKET_NOOP;
+		}
+
+		if (peer->t_refresh_stalepath) {
+			zlog_info(
+				"%s: rcvd route-refresh (BoRR) for %s/%s, whereas BoRR already received",
+				peer->host, afi2str(afi), safi2str(safi));
+			return BGP_PACKET_NOOP;
+		}
+
+		SET_FLAG(peer->af_sflags[afi][safi], PEER_STATUS_BORR_RECEIVED);
+		UNSET_FLAG(peer->af_sflags[afi][safi],
+			   PEER_STATUS_EORR_RECEIVED);
+
+		/* When a BGP speaker receives a BoRR message from
+		 * a peer, it MUST mark all the routes with the given
+		 * Address Family Identifier and Subsequent Address
+		 * Family Identifier, <AFI, SAFI> [RFC2918], from
+		 * that peer as stale.
+		 */
+		if (peer_active_nego(peer)) {
+			SET_FLAG(peer->af_sflags[afi][safi],
+				 PEER_STATUS_ENHANCED_REFRESH);
+			bgp_set_stale_route(peer, afi, safi);
+		}
+
+		if (peer->status == Established)
+			thread_add_timer(bm->master,
+					 bgp_refresh_stalepath_timer_expire,
+					 paf, peer->bgp->stalepath_time,
+					 &peer->t_refresh_stalepath);
+
```

Comment:
should not we host the thread to a struct per afi/safi instead of a per peer structure ?

Reply:
Actually I pass struct peer_af to a thread or what do you mean per peer structure?

Reply:
if you send refresh for multiple afi/Safi, then you will have multiple threads. ( clear bgp <afi> <safi> ...)
If suddenly, the BGP is deleted, you will want to perform THREAD_OFF.
peer->t_refresh_stalepath contains only one thread.

I assume you can not free pending threads

Reply:
Thanks, got it. Will take care of this.

Reply:
@pguibert6WIND looking more into the code seems that graceful restart stalepath timer is also per peer, not per afi/safi. How should we proceed? I think it would be better to sticking how GR is implemented to avoid complexity for now. What do you think?

Reply:
agree. having global timer is ok for me.